[RUM-11519] Add RUM Action Tracking configuration options to the Babel Plugin

cdn34dd opened the pull request:
### What does this PR do?

Adds support for configuration options in the babel plugin targeting RUM Action Tracking.

```
@datadog/mobile-react-native-babel-plugin
└─ options
   ├─ actionNameAttribute?        (string)
   └─ components?
      ├─ useContent?              (boolean)  # global default
      ├─ useNamePrefix?           (boolean)  # global default
      └─ tracked                  (array)
         ├─ [0]
         │  ├─ name               (string)   # component name
         │  ├─ useContent?        (boolean)  # override
         │  ├─ useNamePrefix?     (boolean)  # override
         │  ├─ contentProp?       (string)   # e.g. "subTitle"
         │  └─ handlers           (array)
         │     ├─ [0] { event, action }
         │     └─ [n] { event, action }
         ├─ [1] ...
         └─ [n] ...
```

## Configuration options:

• **actionNameAttribute**
A custom prop you can add to components in your JSX. If present, the plugin uses its value as the action name.
Example:

```jsx

<Button custom-title="Checkout" onPress={...} />

```

• **useContent**
Whether the plugin should try to extract the text content of a component as its action name.
Can be set globally and overridden at the component level.

• **useNamePrefix**
Whether tracked action names should be prefixed with the component’s name.
Example: TextInput("Address") → "Address"
Can be set globally and overridden per component.

• **contentProp**
A property the plugin should read to derive the action name.
By default, the plugin already checks:
    • The component’s children
    • Common props like title or label
With contentProp, you can explicitly tell the plugin which property holds the content.
Example: contentProp: "subTitle"


Example configurations:

```js
module.exports = {
  presets: ['module:@react-native/babel-preset'],
  plugins: [
    [
      '@datadog/mobile-react-native-babel-plugin',
      {
        components: {
          useContent: true,
          useNamePrefix: true,
          tracked: [
            {
              name: 'FancyButton',
              contentProp: 'text'
              handlers: [{event: 'onPress', action: 'TAP'}],
            },
            {
              name: 'CustomTextInput',
              handlers: [{event: 'onFocus', action: 'TAP'}],
            },
            {
              useNamePrefix: false,
              useContent: false,
              name: 'Tab',
              handlers: [{event: 'onChange', action: 'TAP'}],
            },
          ],
        },
      },
    ],
  ],
};
```

### Motivation

What inspired you to submit this pull request?

Users' setups can get very complex and sometimes the base functionality we offer through the babel plugin is not enough. As such we want to give users more control over what gets instrumented and how.

In practice this will allows to support component from third-party libraries or even custom components from a user's codebase.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)
- [ ] If this PR is auto-generated, please make sure also to manually update the code related to the change

cdn34dd committed:
Prevent node_modules files from being affected the babel plugin

cdn34dd committed:
Ensure babel plugin is initialized once per platform

The goal is to prevent caching issue in metro bundler, that would occur
when running both android and iOS apps on the same metro bundler.

cdn34dd committed:
Update babel plugin types & constants to accommodate the new RUM Action Tracking config options

marco-saia-datadog commented on packages/core/src/rum/instrumentation/interactionTracking/DdBabelInteractionTracking.ts:
```diff
@@ -26,10 +26,13 @@ type BabelConfig = {
+    getContent: (() => string[]) | undefined;
+    options: { useContent: boolean; useNamePrefix: boolean };
+    handlerArgs: any[];
     compoenentName: string;
```
typo, it should be `componentName`

marco-saia-datadog commented on packages/core/src/rum/instrumentation/interactionTracking/DdBabelInteractionTracking.ts:
```diff
@@ -64,28 +67,55 @@ export class DdBabelInteractionTracking {
-        return componentName;
+        if (!selectedContent) {
+            return componentName;
+        }
```
1. Isn't it better to invert the order of the if statements, and use else if(s) to determine the `selectedContent`? As in:

```ts
if (useAccessibilityLabel && accessibilityLabel) {
	selectedContent = accessibilityLabel;
} else if (Object.keys(attrs)) {
    selectedContent = attrs[Object.keys(attrs)[0]];
} else if (actionName) {
	selectedContent = actionName;
} else if (content && content.length) {
	selectedContent = content;
} else {
	return componentName;
}
```

2. Can you walk me through the logic of using `Object.keys` on attrs?  It does not seem safe.

cdn34dd replied:
1) No, I don't think it is, as that's not the correct order. The correct order is content (if set to true), then actionName, actionNameAttribute, then accessibilityLabel. Now that I look at it, the code is not correct and does not do what I describe.

2) Should be safe as `attrs` will always defined and will always be an object. And the idea behind this `attrs` value is to handle the `actionNameAttribute` option.
Since this `actionNameAttribute` option can be any value we can't know in our code what it will be, so whenever the plugin finds the config value the user specified at build time it sets it onto this `attrs` object. If not set then we get an empty object.
The reason for the `Object.keys`, is to be able to get the `actionNameAttribute` key.

sbarrio commented on packages/core/src/rum/instrumentation/interactionTracking/DdBabelInteractionTracking.ts:
```diff
@@ -64,28 +67,55 @@ export class DdBabelInteractionTracking {
         } = targetObject;
 
         const { useAccessibilityLabel } = DdBabelInteractionTracking.config;
+        let selectedContent: string[] | null = null;
```
Nit: I get the feeling that this function could be optimized a bit and do early returns by checking the available parameters first and thus avoiding doing extra work that won't be finally used in the end.

cdn34dd replied:
This function is not doing pretty much any processing at all, it just picks the right value, which was already pre-calculated. Not sure how much more optimized you could get it, maybe make it not run the if conditions if one is already picked, but you can't exit early.

marco-saia-datadog commented on packages/core/src/rum/instrumentation/interactionTracking/DdBabelInteractionTracking.ts:
```diff
@@ -64,28 +67,55 @@ export class DdBabelInteractionTracking {
+
+        const content = getContent?.();
+
+        const index = handlerArgs
+            ? handlerArgs.find(x => typeof x === 'number') || 0
+            : 0;
```
Can you explain this logic?

cdn34dd replied:
The goal here is that when you have compound components (usually things like tabs, radio buttons, and so on) we will get multiple potential content values, and usually all those share a value which to match to an index of the item the user clicked.

Example:

```
  <Tab
    value={index}
    onChange={setIndex}
    indicatorStyle={styles.tabIndicator}
    variant="primary">
    <Tab.Item title="Recent" titleStyle={styles.tabTitle} />
    <Tab.Item title="favorite" titleStyle={styles.tabTitle} />
    <Tab.Item title="cart" titleStyle={styles.tabTitle} />
  </Tab>

```

So we try to look for it, if we find a potential index value we try to use that otherwise we default to 0.

It can happen that the value we find is not an actual index, but something else entirely, and in this case we have a fail-safe condition at the end of the function to address that. 

But this is not 100% guaranteed to always get the right content, but should get it in most cases. For the other ones the user needs to configure it in the options.

marco-saia-datadog commented on packages/core/src/rum/instrumentation/interactionTracking/ddBabelUtils.ts:
```diff
@@ -0,0 +1,129 @@
+
+const LABEL_PROPS = ['children', 'label', 'title', 'text'];
+
+const norm = (s: string) => s.replace(/\s+/g, ' ').trim();
```
nit: I would go for `normalize` for the name, as we usually avoid abbreviating function and variable names when possible, and perhaps add a comment along the lines of:

```
/// Normalizes the given string by collapsing multiple spaces into one
```

marco-saia-datadog commented on packages/core/src/rum/instrumentation/interactionTracking/ddBabelUtils.ts:
```diff
@@ -0,0 +1,129 @@
+                    return got;
+                }
+            }
+        }
```
nit: avoid abbreviating variable names.

- `p` could be `prop` or `propKey`
- `v` could be `value`or `propValue`
- `got` could be `text` or `extractedText`

marco-saia-datadog commented on packages/core/src/rum/instrumentation/interactionTracking/ddBabelUtils.ts:
```diff
@@ -0,0 +1,129 @@
+        const kids = React.Children.toArray(props.children);
+        if (kids.length === 0) {
+            return [];
+        }
```
nit: should be `children`

marco-saia-datadog commented on packages/core/src/rum/instrumentation/interactionTracking/ddBabelUtils.ts:
```diff
@@ -0,0 +1,129 @@
+            if (hasLabelProp || childTextCount > 0) {
+                labelledChildCount++;
+            }
+        });
```
- **nit**: avoid abbreviating the variable names
- **ultra-nit**: american english is usually preferred for variable names (`labeled` instead of `labelled`)

sbarrio commented on packages/react-native-babel-plugin/src/actions/rum/tap.ts:
```diff
@@ -23,20 +32,23 @@ export function handleTapAction(
-        t.isIdentifier(expression) ||
-        (t.isMemberExpression(expression) &&
-            t.isThisExpression(expression.object));
+        t.isIdentifier(expression) || t.isMemberExpression(expression); // && t.isThisExpression(expression.object));
```
Leftover?

marco-saia-datadog replied:
nit: I would either make the comment more descriptive by explaining what it does, or remove it if it is not needed anymore.

sbarrio commented on packages/react-native-babel-plugin/src/actions/rum/tap.ts:
```diff
@@ -76,18 +99,35 @@ export function handleTapAction(
-          )
-        : t.callExpression(expression, [t.spreadElement(t.identifier('args'))]);
+
+    // const returnExpression = isArrowFunc
```
Another leftover?

marco-saia-datadog commented on packages/core/src/rum/instrumentation/interactionTracking/ddBabelUtils.ts:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    // Zero-arg render prop
+    if (typeof node === 'function' && node.length === 0) {
+        try {
+            return __ddExtractText(node());
+        } catch {
+            return [];
+        }
+    }
```
I don't think it is safe to blindly call the function. What use cases does this cover? Can we tailor the implementation to whitelist only those?

cdn34dd replied:
The goal here is to handle scenarios like the one below. While not very common can still happen. 

Now regarding if it's safe, given the context, I'd say it is, we're looking for a very specific pattern, where the children is a function of length 0 (no attributes), which is also a common pattern.

And we're retrieving the contents of elements like buttons, we're not getting the contents of really complex components, so I don't really see to much reason for concern. If we were targeting more complex elements, then this function execution could be doing something heavy, and that would not be great.

```
type Props = {
  onPress: () => void;
  children: () => React.ReactNode;
};

export const CustomButton = ({onPress, children}: Props) => {
  const content = children();

  return (
    <TouchableOpacity onPress={onPress}>
      <View style={{padding: 12, borderRadius: 8, borderWidth: 1}}>
        {content}
      </View>
    </TouchableOpacity>
  );
};
```

```
  <CustomButton onPress={() => console.log('Pressed!')}>
    {() => <Text>Button Text</Text>}
  </CustomButton>
 ```

marco-saia-datadog commented on packages/react-native-babel-plugin/src/actions/global/index.ts:
```diff
@@ -52,28 +54,33 @@ export function loadImportMap(
+                        useNamePrefix: options.components.useNamePrefix,
+                        handlers: elementEvents.map(event => ({
+                            event,
+                            action: RumAction.TAP // TODO: change once we support more actions
```
If we have to leave TODOs in the codebase, let's add a JIRA ticket number as reference in the comment, so that we can keep track of them

cdn34dd replied:
The main issue with that, is that this will likely be many jira tickets, as each action will likely be one. And currently no tickets exist regarding these changes.

marco-saia-datadog commented on packages/react-native-babel-plugin/src/actions/rum/tap.ts:
```diff
@@ -23,20 +32,23 @@ export function handleTapAction(
 
-    // Check if the property and element is valid by checking our tap mappings list
-    const mapEntry = state.tapMappings?.[parentName];
+    // Check if the property and element is valid by checking our tap mappings list interaction handlers for RUM should wrap arrow
```
I think this is here by mistake: `interaction handlers for RUM should wrap arrow`

marco-saia-datadog commented on packages/react-native-babel-plugin/src/actions/rum/tap.ts:
```diff
@@ -76,18 +99,35 @@ export function handleTapAction(
-              getArgumentsFromParams(t, state, expression.params).callArgs
-          )
-        : t.callExpression(expression, [t.spreadElement(t.identifier('args'))]);
+
+    // const returnExpression = isArrowFunc
+    //     ? t.callExpression(
+    //           expression,
+    //           getArgumentsFromParams(t, state, expression.params).callArgs
+    //       )
+    //     : t.callExpression(expression, [t.spreadElement(t.identifier('args'))]);
+
```
I think this should be removed, seems like a copy-paste of the code below

marco-saia-datadog commented on packages/react-native-babel-plugin/src/actions/rum/tap.ts:
```diff
@@ -351,6 +464,29 @@ function getActionWrapperFunction(
+    const wrapperExpressionImmediate = t.callExpression(
+        wrapperExpression,
+        callArgs
+        // [t.spreadElement(t.identifier('args'))]
```
nit: I would either remove this comment, or make it more descriptive

marco-saia-datadog commented on packages/react-native-babel-plugin/src/index.ts:
```diff
@@ -21,24 +22,54 @@ import type {
+
                         const { path: p, name } = getFileInfo(this);
 
+                        if (p?.includes('node_modules')) {
+                            return;
+                        }
```
Were node modules also affected by our plugin before this change?

cdn34dd replied:
Yes, weirdly enough it was, but only certain modules (maybe related to some metro configuration), that could potentially cause issues with our code, and was not the intended goal in the first place.

marco-saia-datadog reviewed (commented):
Overall, this looks good, great work 🚀  

There are just a few points I'd like to discuss (mainly [this one](https://github.com/DataDog/dd-sdk-reactnative/pull/968#discussion_r2316182583)) before approval, and I have left some minor suggestions.

cdn34dd committed:
Add Babel config options for RUM Action Tracking & support content extraction

- Allow customization through Babel options to extend RUM Action
Tracking to custom/third-party components
- Allow for retrieval of components' content through children or props

cdn34dd committed:
Update Babel Plugin's unit tests to accommodate for plugin options

cdn34dd committed:
Update Babel plugin's README.md to include new configuration options

rtrieu commented on packages/react-native-babel-plugin/README.md:
```diff
@@ -22,7 +22,7 @@ yarn add @datadog/mobile-react-native-babel-plugin
 Add the plugin to your Babel configuration. Depending on your setup, you might be using a `babel.config.js`, `.babelrc`, or similar.
 
-Example configuration:
+<b>Example configuration:</b>
```
```suggestion
**Example configuration:**
```

rtrieu commented on packages/react-native-babel-plugin/README.md:
```diff
@@ -31,7 +31,53 @@ module.exports = {
+
+| Option          | Type    | Default | Description |
+|-----------------|---------|---------|-------------|
+| `useContent`    | boolean | true   | Whether to use component content (e.g., children, props) as the action name. |
```
```suggestion
| `useContent`    | boolean | true   | Whether to use component content (for example: children, props) as the action name. |
```

rtrieu commented on packages/react-native-babel-plugin/README.md:
```diff
@@ -31,7 +31,53 @@ module.exports = {
+| `name`          | string  | –                    | The component name to track (e.g., `Button`). |
+| `useContent`    | boolean | inherits from global | Override `useContent` for this component. |
+| `useNamePrefix` | boolean | inherits from global | Override `useNamePrefix` for this component. |
+| `contentProp`   | string  | –                    | Property name to use for content instead of children (e.g., `"subTitle"`). |
```
```suggestion
| `contentProp`   | string  | –                    | Property name to use for content instead of children (for example, `"subTitle"`). |
```

rtrieu commented on packages/react-native-babel-plugin/README.md:
```diff
@@ -31,7 +31,53 @@ module.exports = {
+
+| Field   | Type   | Description |
+|---------|--------|-------------|
+| `event` | string | The event name to intercept (e.g., `"onPress"`). |
```
```suggestion
| `event` | string | The event name to intercept (such as `"onPress"`). |
```

rtrieu commented on packages/react-native-babel-plugin/README.md:
```diff
@@ -31,7 +31,53 @@ module.exports = {
+| Field   | Type   | Description |
+|---------|--------|-------------|
+| `event` | string | The event name to intercept (e.g., `"onPress"`). |
+| `action`| string | The RUM action name to associate with this event. <i>(Only `"TAP"` actions are currently supported)<i> |
```
```suggestion
| `action`| string | The RUM action name to associate with this event. _(Only `"TAP"` actions are currently supported)_ |
```

rtrieu commented on packages/react-native-babel-plugin/README.md:
```diff
@@ -31,7 +31,53 @@ module.exports = {
+
+---
+
+<b>Example configuration (<i>using configuration options</i>):</b>
```
```suggestion
**Example configuration (_using configuration options_):**
```

rtrieu reviewed (dismissed):
some minor feedback, mostly suggesting markdown vs html
